File: src/java/org/algoritmos/ordenamiento/RadixSort.java
package org.algoritmos.ordenamiento;

/**
 * Radix Sort (Base 10)
 *
 * Algoritmo de ordenamiento no comparativo que ordena digito por digito.
 *
 * Fuente: GeeksforGeeks
 *
 * Complejidad: O(d * (n + k))
 */
public class RadixSort {

    private static final int BASE = 10;

    /**
     * Ordena el arreglo usando Radix Sort.
     *
     * @param array Arreglo de enteros positivos a ordenar
     * @return El mismo arreglo ordenado
     */
    public static int[] sort(int[] array) {
        if (array == null || array.length < 2) {
            return array;
        }

        int max = findMax(array);

        for (int exp = 1; max / exp > 0; exp *= BASE) {
            countingSortByDigit(array, exp);
        }

        return array;
    }

    private static int findMax(int[] array) {
        int max = array[0];
        for (int value : array) {
            if (value > max) {
                max = value;
            }
        }
        return max;
    }

    private static void countingSortByDigit(int[] array, int exp) {
        int n = array.length;
        int[] output = new int[n];
        int[] count = new int[BASE];

        for (int value : array) {
            int digit = (value / exp) % BASE;
            count[digit]++;
        }

        for (int i = 1; i < BASE; i++) {
            count[i] += count[i - 1];
        }

        for (int i = n - 1; i >= 0; i--) {
            int digit = (array[i] / exp) % BASE;
            output[count[digit] - 1] = array[i];
            count[digit]--;
        }

        System.arraycopy(output, 0, array, 0, n);
    }
}

File: src/python/ordenamiento/radix_sort.py
"""
Radix Sort - Implementacion en Python (base 10).
Fuente: GeeksforGeeks

Algoritmo de ordenamiento no comparativo que ordena enteros
procesando digito por digito.

Complejidad: O(d * (n + k)) donde d = digitos, n = elementos, k = base
"""

from typing import List


def radix_sort(arr: List[int]) -> List[int]:
    """
    Ordena un arreglo de enteros no negativos usando Radix Sort.

    Args:
        arr: Lista de enteros no negativos a ordenar

    Returns:
        Nueva lista ordenada en orden ascendente

    Raises:
        ValueError: Si hay numeros negativos en el arreglo
    """
    if not arr:
        return []

    result = arr.copy()
    n = len(result)

    if n <= 1:
        return result

    # Verificar que no haya numeros negativos
    if any(x < 0 for x in result):
        raise ValueError("Radix Sort no soporta numeros negativos.")

    max_num = max(result)

    if max_num == 0:
        return result

    exp = 1
    while max_num // exp > 0:
        _counting_sort_by_digit(result, exp)
        exp *= 10

    return result


def _counting_sort_by_digit(arr: List[int], exp: int) -> None:
    """Counting Sort para ordenar por un digito especifico."""
    n = len(arr)
    output = [0] * n
    count = [0] * 10

    for i in range(n):
        digit = (arr[i] // exp) % 10
        count[digit] += 1

    for i in range(1, 10):
        count[i] += count[i - 1]

    for i in range(n - 1, -1, -1):
        digit = (arr[i] // exp) % 10
        output[count[digit] - 1] = arr[i]
        count[digit] -= 1

    for i in range(n):
        arr[i] = output[i]
